### ISSUE TITLE
[RFC] Implement TypedSettings

### ISSUE THREAD

**Issue by rochacbruno** [dynaconf/dynaconf#1104]
```python

#### Internal Dynaconf ####

from typing import Annotated, cast, TYPE_CHECKING, reveal_type
from dataclasses import dataclass
from dynaconf import Validator, Dynaconf


class TypedSettings:
    @classmethod
    def build(cls, **options):
        # Extract validators from cls.__annotations__[x].__metadata__
        return cast(cls, Dynaconf(**options))

#### Usage

from dynaconf import TypedSettings, Validator


@dataclass
class MySchema(TypedSettings):
    name: Annotated[str, Validator(ne="test")]
    port: int


# builder pattern
settings: MySchema = MySchema.build(name="test")


```

**Comment by pedro-psb** [comment]
Some notes:

**What kind of Schema declarations we'll support?**

- On our discussion I've said we should only support the `Annotated` initially, but I think we should support simple types casting too (`: int`, `: Path`, `: str`).
  - The reason is that this is very common. I would mostly use that on `pulp-docs`, for example.
  - In the example, `port: int` would generate a Validator that try to cast the value to int.
  - Typed containers (`key: list[int]`) may be a bit harder to implement. For the sake of getting things done we may left those out if they become blockers.

**How the user is going to use the schema?**

- The only thing that bothers me about `MySchema.build(name="test")` is that we'll add another reserved word (`build`).
  - It wouldn't really prevent the user from using a  `build` key on their settings, since this exist only on the TypeSettings "stub", so its not that bad.
  -  TLDR, I'm ok with this design that we have agreed. I'll just add throw my preference here, case you take this and change  your mind. 

```python
# new typed module, so Dynaconf here is just a factory function
from dynaconf.typed import Dynaconf, Validator, TypedSettings

@dataclass
class MySchema(TypedSettings): ...

settings: MySchema = Dynaconf(MySchema, name="test")
```
**What about disallowing keys not defined in the schema?**

- If that is easy enough to implement, we should add it now. If its blocking the feature, lets do it later.
  - I change my mind often. Looking at this now, it really feels the schema default should be disallowing "un-declared" keys.
  - Even so, lets not be blocked by this.

**Other observations:**

- We should also test for nested Schemas cases, but I don't think that's going to be a problem. I guess the approach is to recurse case the key type is a subclass of `TypedSettings`.

Also, related to #1082

**Comment by rochacbruno** [comment]
@pedro-psb what I have is:

save on an `app.py`

```python
# ---------------- dynaconf.typed -----------------------------

from typing import TYPE_CHECKING, cast, Annotated
from dynaconf.base import Settings as BaseSettings
from dynaconf import Dynaconf as OriginalDynaconf
from dynaconf import Validator as Validator


class Dynaconf(BaseSettings):
    def __new__(cls, *args, **kwargs):
        validators = kwargs.pop("validators", [])
        for name, annotation in cls.__annotations__.items():
            if _validators := getattr(annotation, "__metadata__", None):
                for _validator in _validators:
                    _validator.names = [name]
                    validators.append(_validator)
            else:
                validators.append(Validator(name, is_type_of=annotation))

        # Eager validation
        if validators:
            kwargs["validators"] = validators

        new_cls = OriginalDynaconf(*args, **kwargs)
        new_cls.__annotations__ = cls.__annotations__

        # Lazy validation
        # if validators:
        #    new_cls.validators.register(**validators)

        return cast(cls, new_cls)

# ------------------ env ---------------------

import os
os.environ["MYAPP_PORT"] = "1234"
# os.environ["MYAPP_PORT"] = "umdoistresquatro"  # will raise validation Error
# os.environ["MYAPP_HOST"] = "batata.com"  # will raise Validation Error

os.environ["MYAPP_THING"] = "value"  # just loaded even if not part of the schema

# ------------------ app.py ------------------

#from dynaconf.typed import Dynaconf, Validator
from typing import Annotated

# No need to be a dataclass, we never instantiate it, we just use the annotations
class Settings(Dynaconf):
    host: Annotated[str, Validator(ne="batata.com")]
    port: int


settings = Settings(envvar_prefix="MYAPP", host="server.com")

# part of the schema, loaded and validated
assert settings.host == "server.com"
assert settings.port == 1234

# out of schema but loaded anyway
# we can think of an option to ignore undefined values like this in the env
# or to raise an exception in the case of a value like this on files
assert settings.thing == "value"

```

**Comment by pedro-psb** [comment]
Looks great, I believe that's the design we want. Copying notes from matrix:

> I've tested here as well and it worked on helix and vscode
> I've noticed on python 3.12 it mixes dynaconf and the schema, while in 3.9 it only get schema data. But I wouldn't worry about that now, the most important data is the schema, and we can do the stub later if we really need to

**Comment by jtpavlock** [comment]
Is 0cdc5e7 planned for an upcoming release?

**Comment by pedro-psb** [comment]
Yes. We are trying to get these done to make the release, but we are a bit low on capacity:
https://github.com/dynaconf/dynaconf/issues?q=is%3Aissue%20state%3Aopen%20milestone%3A3.3.0
